srcjs/output_binding_html.js

```diff
@@ -107,23 +107,55 @@ function renderDependency(dep) {
-      var re = new RegExp(this_href + "$");
-      for (var i = 0; i < document.styleSheets.length; i++) {
-        var h = document.styleSheets[i].href;
-        if (!h) continue;
-        if (restyle && h.match(re)) document.styleSheets[i].disabled = true;
-      }
-      // Force client to re-request the stylesheet if we've already seen it
-      // (without this unique param, the request might get cached)
-      if (restyle) {
-        this_href = this_href + "?restyle=" + new Date().getTime();
-      }
-      return $("<link rel='stylesheet' type='text/css'>").attr("href", this_href);
+    var links = $.map(asArray(dep.stylesheet), function(stylesheet) {
+      return $("<link rel='stylesheet' type='text/css'>").attr("href", href + "/" + encodeURI(stylesheet));
     });
-    $head.append(stylesheets);
+
+    if (!restyle) {
+      $head.append(links);
+    } else {
+      $.map(links, function(link) {
+        // Flag 'stale' stylesheets for disable+removal after new styles are inserted
+        for (var i = 0; i < document.styleSheets.length; i++) {
+          var sheet = document.styleSheets[i];
+          if (typeof sheet.href !== "string") continue;
+          // The sheet's href is a full URL
+          if (sheet.href.indexOf(link.attr("href")) === -1) continue;
+          sheet._shiny_restyle = true;
+        }
+        // Request the new stylesheet (with a timestamp to prevent caching)
+        // and insert a new inline style. Note that this could also done by inserting
+        // a the new <link> (w/ timestamp in href), then disabling the 'stale'
+        // document.styleSheets, but that approach doesn't work with IE11 (and this one does)
+        var xhr = new XMLHttpRequest();
+        var url = link.attr("href") + "?restyle=" + new Date().getTime();
+        xhr.open('GET', url);
+        xhr.onreadystatechange = function() {
+          if (this.readyState === 4 && this.status === 200) {
+            xhr.onreadystatechange = null;
+            var id = "shiny_restyle_" + link.attr("href").replace(/\W/g,'_');
+            var style_old = $head.find("style#" + id);
+            var style_new = $("<style>").attr("id", id).html(this.responseText);
+            $head.append(style_new);
+            style_old.remove();
+          }
+        };
+        xhr.send();
+      });
+      // Disable & remove 'stale' <link> tags on the next tick (to avoid FOUC)
+      setTimeout(function() {
+        var staleNodes = $.map(document.styleSheets, function(sheet) {
+          if (!sheet._shiny_restyle) return null;
+          sheet.disabled = true;
+          // disabled=true doesn't work on IE11, but this does :shrug:?
+          if (navigator.appName === 'Microsoft Internet Explorer') {
+            sheet.cssText = "";
+          }
+          return sheet.ownerNode;
+        });
+        $.map(staleNodes, function(node) { if (node) $(node).remove(); });
+      }, 100);
```

Comment:
This assumes that all the stylesheets would have loaded within 100ms, which may not be true, especially on slow connections.